pkg/metrics/prometheus.go

```diff
@@ -0,0 +1,152 @@
+package metrics
+
+import (
+	"github.com/grafana/grafana/pkg/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"net/http"
+	"strings"
+)
+
+const (
+	namespace = "grafana"
+	subsystem = "backend"
+)
+
+type PrometheusPublisher struct {
+	CounterCollectors map[string]*prometheus.CounterVec
+	GaugeCollectors   map[string]*prometheus.GaugeVec
+}
+
+func CreatePrometheusPublisher() (*PrometheusPublisher, error) {
+	go func() {
+		http.Handle("/metrics", prometheus.Handler())
+		err := http.ListenAndServe(":9200", nil)
```

Comment:
> So prometheus needs its golang client library embedded in all servers/apps it wants to monitor?

It doesn't, but if you want to support multiple different monitoring systems it's a good choice.

> Yes, and storing counters as deltas makes analytics & rollups so much easier and powerful.

I wouldn't agree there. It's similar complexity math both ways.

> You store timers as two counts? The timers in Grafana provide these metrics:

That's not particularly useful data. Those look like Dropwizard metrics, of those we can only use the "count" and we can't calculate latency from those as there's no matching "sum". The rest aren't of much use statistically (https://prometheus.io/docs/instrumenting/writing_exporters/#drop-less-useful-statistics).

> working with rate functions that you then have to multiply to get bigger time periods or sum is very inaccurate or complex and usually breaks down when you hit roll ups periods (not a problem for Prometheus I guess as it does not support rollups). 

That's not a problem with rates. That's a fundamental problem with the raw data.

> Storing deltas makes analytics so much easier, you can just sum them, and the rollups are summed so no matter what time range you can get accurate sums.

That works just as well with counters, and I'd expect Prometheus to produce a slightly more accurate answer due to some additional math we do in our `rate()` function.

> And super easy to turn them into rates with graphites scaleToSeconds (which knows how many seconds each value represents so it's "accurate" no matter what rollup you hit).

Last I checked Graphite was not accurate in this regard, as it can't be. Similarly Prometheus isn't accurate.